File: python-sdk/examples/error_handling.py
"""
Error handling example for LLM-CostOps SDK.

This example demonstrates proper error handling for various scenarios.
"""

from llm_cost_ops import CostOpsClient
from llm_cost_ops.exceptions import (
    CostOpsError,
    APIError,
    AuthenticationError,
    AuthorizationError,
    ValidationError,
    NotFoundError,
    RateLimitError,
    ServerError,
    NetworkError,
    TimeoutError,
    ConfigurationError,
)
import time


def handle_specific_errors() -> None:
    """Demonstrate handling specific error types."""
    print("=== Handling Specific Errors ===")

    try:
        # This will fail with invalid API key
        client = CostOpsClient(api_key="invalid-key")
        client.usage.submit(
            organization_id="org-123",
            provider="openai",
            model_id="gpt-4",
            input_tokens=1000,
            output_tokens=500,
            total_tokens=1500,
        )
    except AuthenticationError as e:
        print(f"Authentication failed: {e.message}")
        print(f"Status code: {e.status_code}")
        print(f"Request ID: {e.request_id}")
    except ValidationError as e:
        print(f"Validation error: {e.message}")
        print(f"Errors: {e.errors}")
    except RateLimitError as e:
        print(f"Rate limit exceeded: {e.message}")
        print(f"Retry after: {e.retry_after} seconds")
        print(f"Limit: {e.limit} requests")
        # Wait and retry
        if e.retry_after:
            print(f"Waiting {e.retry_after} seconds before retry...")
            time.sleep(e.retry_after)
    except NotFoundError as e:
        print(f"Resource not found: {e.message}")
        print(f"Resource type: {e.resource_type}")
    except ServerError as e:
        print(f"Server error: {e.message}")
        print(f"Status code: {e.status_code}")
        # Implement exponential backoff for server errors
    except NetworkError as e:
        print(f"Network error: {e.message}")
        print(f"Cause: {e.cause}")
        # Check network connection and retry
    except TimeoutError as e:
        print(f"Request timed out: {e.message}")
        print(f"Timeout: {e.timeout} seconds")
        # Increase timeout or retry
    except APIError as e:
        # Catch-all for other API errors
        print(f"API error: {e.message}")
        print(f"Status code: {e.status_code}")
    except ConfigurationError as e:
        print(f"Configuration error: {e.message}")
        print(f"Parameter: {e.parameter}")
    except CostOpsError as e:
        # Base exception - catch-all
        print(f"General SDK error: {e.message}")
    finally:
        print()


def handle_with_retry() -> None:
    """Demonstrate error handling with retry logic."""
    print("=== Error Handling with Retry ===")

    max_retries = 3
    retry_delay = 1.0

    for attempt in range(max_retries):
        try:
            client = CostOpsClient(api_key="your-api-key-here")
            usage = client.usage.submit(
                organization_id="org-123",
                provider="openai",
                model_id="gpt-4",
                input_tokens=1000,
                output_tokens=500,
                total_tokens=1500,
            )
            print(f"Success! Usage ID: {usage.usage_id}")
            client.close()
            break

        except (ServerError, NetworkError, TimeoutError) as e:
            print(f"Attempt {attempt + 1} failed: {e.message}")

            if attempt < max_retries - 1:
                wait_time = retry_delay * (2**attempt)  # Exponential backoff
                print(f"Retrying in {wait_time} seconds...")
                time.sleep(wait_time)
            else:
                print("Max retries reached. Giving up.")
                raise

        except RateLimitError as e:
            print(f"Rate limited: {e.message}")
            if e.retry_after and attempt < max_retries - 1:
                print(f"Waiting {e.retry_after} seconds as instructed...")
                time.sleep(e.retry_after)
            else:
                print("Cannot retry rate limit.")
                raise

        except (AuthenticationError, ValidationError) as e:
            # Don't retry authentication or validation errors
            print(f"Non-retryable error: {e.message}")
            raise

    print()


def main() -> None:
    """Run error handling examples."""
    handle_specific_errors()

    # Uncomment to test retry logic
    # handle_with_retry()


if __name__ == "__main__":
    main()

File: sdk/examples/error-handling.ts
/**
 * Error handling example
 */

import {
  CostOpsClient,
  AuthenticationError,
  ValidationError,
  RateLimitError,
  NotFoundError,
  isCostOpsError,
} from '../src/index.js';

async function main(): Promise<void> {
  const client = new CostOpsClient({
    baseUrl: 'https://api.llm-cost-ops.example.com',
    apiKey: process.env['API_KEY'] ?? 'your-api-key',
  });

  // Handle specific error types
  try {
    await client.getMetric('non-existent-id');
  } catch (error) {
    if (error instanceof NotFoundError) {
      console.log('Metric not found:', error.message);
      console.log('Resource ID:', error.resourceId);
    } else if (error instanceof AuthenticationError) {
      console.error('Authentication failed - check your API key');
    } else if (error instanceof ValidationError) {
      console.error('Validation error:', {
        field: error.field,
        message: error.message,
        constraints: error.constraints,
      });
    } else if (error instanceof RateLimitError) {
      console.error('Rate limited - retry after:', error.retryAfter, 'seconds');
    } else if (isCostOpsError(error)) {
      console.error('SDK Error:', {
        name: error.name,
        message: error.message,
        code: error.code,
        timestamp: error.timestamp,
      });
    } else {
      console.error('Unknown error:', error);
    }
  }

  // Handle validation errors
  try {
    await client.createMetric({
      service: '', // Invalid: empty string
      cost: -10, // Invalid: negative cost
      currency: 'USD',
      timestamp: new Date().toISOString(),
    });
  } catch (error) {
    if (error instanceof ValidationError) {
      console.error('Validation failed:', error.field);
    }
  }

  // Handle rate limiting
  try {
    // Make many requests in quick succession
    const promises = Array.from({ length: 100 }, (_, i) =>
      client.getMetrics({ limit: 1, offset: i })
    );
    await Promise.all(promises);
  } catch (error) {
    if (error instanceof RateLimitError) {
      console.log('Rate limited!');
      console.log('Retry after:', error.retryAfter, 'seconds');
      console.log('Rate limit:', error.limit, 'requests');
    }
  }

  // Graceful error handling with fallbacks
  async function getMetricWithFallback(id: string): Promise<any | null> {
    try {
      return await client.getMetric(id);
    } catch (error) {
      if (error instanceof NotFoundError) {
        console.warn(`Metric ${id} not found, returning null`);
        return null;
      }
      throw error; // Re-throw other errors
    }
  }

  const metric = await getMetricWithFallback('test-id');
  console.log('Metric:', metric);
}

main().catch(console.error);
